libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Security/Encryptor.swift

```diff
@@ -0,0 +1,180 @@
+//  endorse or promote products derived from this software without specific prior written
+//  permission of salesforce.com, inc.
+//
+//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR
+//  IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
+//  FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+//  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+//  DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+//  DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+//  WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY
+//  WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import Foundation
+import CryptoKit
+import SalesforceSDKCommon
+
+
+@objc(SFSDKEncryptor)
+public class Encryptor: NSObject {
+    
+    enum EncryptorError: Error {
+        case combinedBoxFailed
+    }
+
+    // MARK: Symmetric Encrypt/Decrypt
+    @objc @available(swift, obsoleted: 1.0) // Objective-c only wrapper
+    public static func encrypt(data: Data, using key: Data) throws -> Data {
+        let symmetricKey = SymmetricKey(data: key)
+        return try encrypt(data: data, using: symmetricKey)
+    }
+
+    public static func encrypt(data: Data, using key: SymmetricKey) throws -> Data {
+        let sealedBox = try AES.GCM.seal(data, using: key, nonce: AES.GCM.Nonce())
+        if let combined = sealedBox.combined {
+            return combined
+        } else {
+            throw EncryptorError.combinedBoxFailed
+        }
+    }
+    
+    @objc @available(swift, obsoleted: 1.0) // Objective-c only wrapper
+    public static func decrypt(data: Data, using key: Data) throws -> Data {
+        let symmetricKey = SymmetricKey(data: key)
+        return try decrypt(data: data, using: symmetricKey)
+    }
+
+    public static func decrypt(data: Data, using key: SymmetricKey) throws -> Data {
+        let sealedBox = try AES.GCM.SealedBox(combined: data)
+        return try AES.GCM.open(sealedBox, using: key)
+    }
+    
+    // MARK: EC Encrypt/Decrypt
+    static func encrypt(data: Data, privateKey: Data) throws -> Data {
+        let key = try derivedKey(from: privateKey)
+        return try encrypt(data: data, using: key)
+    }
+    
+    static func decrypt(data: Data, privateKey: Data) throws -> Data {
+        let key = try derivedKey(from: privateKey)
+        return try decrypt(data: data, using: key)
+    }
+    
+    static func derivedKey(from privateKey: Data) throws -> SymmetricKey {
+        var sharedSecret: SharedSecret
+        if SecureEnclave.isAvailable && !UIDevice.current.isSimulator() {
+            let key = try SecureEnclave.P256.KeyAgreement.PrivateKey(dataRepresentation: privateKey)
```

Comment:
The only way I saw to encrypt with the EC key is to derive a symmetric key from it and then use that, let me know if there are any other API routes I missed

Reply:
The existing [encryptUsingECforData](https://github.com/forcedotcom/SalesforceMobileSDK-iOS/blob/dev/libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Security/SFSDKCryptoUtils.m#L331) is used by [SFSecureEncryptionKey]( https://github.com/forcedotcom/SalesforceMobileSDK-iOS/blob/9f4daa6f28fb1dbeaeee0e611bd550ece66e408e/libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Security/SFSecureEncryptionKey.m#L141) with the private key stored in secure enclave. So the key is never actually in memory, the data to be encrypted/decrypted is sent to that dedicated chip and the resulting data is returned. That makes it very secure. Even on a jailbroken device, you won't to get the key.
If we use a derive key, then that extra protection is lost.

Reply:
thanks! Sounds like we'd only be able to move the symmetric keys to CryptoKit then, I'll experiment with a combo approach